File: apps/php/design-pattern/src/Observer/Employee.php
<?php

declare(strict_types=1);

namespace DesignPattern\Observer;

interface Observer
{
    public function update(Employee $employee): void;
}

class Employee
{
    private string $name;
    private string $title;
    private float $salary;
    private \SplObjectStorage $observers;

    public function __construct(string $name, string $title, float $salary)
    {
        $this->name = $name;
        $this->title = $title;
        $this->salary = $salary;
        $this->observers = new \SplObjectStorage();
    }

    public function getName(): string
    {
        return $this->name;
    }

    public function getTitle(): string
    {
        return $this->title;
    }

    public function getSalary(): float
    {
        return $this->salary;
    }

    public function setSalary(float $salary): void
    {
        $this->salary = $salary;
        $this->notifyObservers();
    }

    public function setTitle(string $title): void
    {
        $this->title = $title;
        $this->notifyObservers();
    }

    public function addObserver(Observer $observer): void
    {
        $this->observers->offsetSet($observer);
    }

    public function removeObserver(Observer $observer): void
    {
        $this->observers->offsetUnset($observer);
    }

    private function notifyObservers(): void
    {
        foreach ($this->observers as $observer) {
            $observer->update($this);
        }
    }
}

File: apps/java/design-pattern/src/main/java/pattern/observer/Employee.java
package pattern.observer;

import java.util.ArrayList;
import java.util.List;

/**
 * 従業員クラス（Observer パターン - Subject）
 *
 * オブザーバーのリストを保持し、給与やタイトルの変更を通知する。
 */
public class Employee {

    private final String name;
    private String title;
    private int salary;
    private final List<Observer> observers = new ArrayList<>();

    public Employee(String name, String title, int salary) {
        this.name = name;
        this.title = title;
        this.salary = salary;
    }

    public void addObserver(Observer observer) {
        observers.add(observer);
    }

    public void removeObserver(Observer observer) {
        observers.remove(observer);
    }

    public void setSalary(int newSalary) {
        this.salary = newSalary;
        notifyObservers();
    }

    public void setTitle(String newTitle) {
        this.title = newTitle;
        notifyObservers();
    }

    private void notifyObservers() {
        for (Observer observer : observers) {
            observer.update(this);
        }
    }

    public String getName() {
        return name;
    }

    public String getTitle() {
        return title;
    }

    public int getSalary() {
        return salary;
    }
}
